### PR TITLE
boost: Clean, reducing duplication

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

This was motivated originally by my cross work, but that goal requires a few more commits to other things. Still, it's good to start the cleanup now / get things out of the way.

###### Things done

Native is almost exactly tested as part of #26805. However, In that version I had this setup hook replicating some logic with `findInputs` downstream that's since been removed. I removed the setup hook for this PR so now this is technically untested, but I doubt anything changed.

Actually its even better, the setup hook traverses propagated dependencies, but no version of boost has any, so its demonstrably a no-op!

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Ericson2314** [comment]
I differed the derivations, and nothing interesting is different.

**Comment by Ericson2314** [comment]
Looking to *finally* merge https://github.com/NixOS/nixpkgs/pull/26805, so Boost maintainer, whoever you are, I hope you aren't on an internet-free vacation! (This blocks that.)

**Review by orivej — APPROVED** [review]

File: pkgs/development/libraries/boost/generic.nix [hunk 1]
```diff
@@ -46,86 +50,53 @@ let
...
-           else if enableExceptions then
-             "cflags=-fexceptions"
-           else
-             "";
+  cflags = concatStringsSep " "
+    (optional (enablePIC) "-fPIC" ++
+     optional (enableExceptions) "-fexceptions");
 
-  withToolset = stdenv.lib.optionalString (toolset != null) "--with-toolset=${toolset}";
+  cxxflags = optionalString (enablePIC) "-fPIC";
```
Comment (globin) [inline pkgs/development/libraries/boost/generic.nix]: You should be able to drop all PIC handling, "fixed" by hardening ages ago

File: pkgs/development/libraries/boost/generic.nix [hunk 2]
```diff
@@ -46,86 +50,53 @@ let
...
 
-  genericB2Flags = [
+  linkflags = optionalString (enablePIC) "-fPIC";
```
Comment (globin) [inline pkgs/development/libraries/boost/generic.nix]: dito

File: pkgs/development/libraries/boost/generic.nix [hunk 3]
```diff
@@ -46,86 +50,53 @@ let
...
-  cflags = if enablePIC && enableExceptions then
-             "cflags=\"-fPIC -fexceptions\" cxxflags=-fPIC linkflags=-fPIC"
-           else if enablePIC then
-             "cflags=-fPIC cxxflags=-fPIC linkflags=-fPIC"
-           else if enableExceptions then
-             "cflags=-fexceptions"
-           else
-             "";
+  cflags = concatStringsSep " "
+    (optional (enablePIC) "-fPIC" ++
```
Comment (globin) [inline pkgs/development/libraries/boost/generic.nix]: dito

**Review by globin — APPROVED** [review]
I've not reviewed this too thoroughly but looks fine generally, you can even probably drop the PIC stuff, see inline comments.

**Comment by Ericson2314** [comment]
@globin Ah, you mentioned that before, and I guess I forgot! turns out the option was unused so it was dead code anyways.

File: pkgs/development/libraries/boost/generic.nix [hunk 4]
```diff
@@ -1,18 +1,19 @@
...
+, toolset ? /**/ if stdenv.cc.isClang                                then "clang"
+            else if stdenv.cc.isGNU && hostPlatform != buildPlatform then "gcc-cross"
+            else null
 , enableRelease ? true
 , enableDebug ? false
 , enableSingleThreaded ? false
 , enableMultiThreaded ? true
 , enableShared ? !(hostPlatform.libc == "msvcrt") # problems for now
 , enableStatic ? !enableShared
-, enablePIC ? false
```
Comment (Ericson2314) [inline pkgs/development/libraries/boost/generic.nix]: This is unused, and hardening provides it anyways

File: pkgs/development/libraries/boost/generic.nix [hunk 5]
```diff
@@ -1,18 +1,19 @@
...
-, enableExceptions ? false
```
Comment (Ericson2314) [inline pkgs/development/libraries/boost/generic.nix]: this is unused

**Pull request merged**